solls/LanguageServer.cpp

```diff
@@ -0,0 +1,679 @@
+			continue;
+
+		auto const message = SourceReferenceExtractor::extract(*error);
+
+		auto const severity = toDiagnosticSeverity(error->type());
+
+		// global warnings don't have positions in the source code - TODO: default them to top of file?
+
+		auto const startPosition = LineColumn{{
+			max(message.primary.position.line, 0),
+			max(message.primary.startColumn, 0)
+		}};
+
+		auto const endPosition = LineColumn{{
+			max(message.primary.position.line, 0),
+			max(message.primary.endColumn, 0)
+		}};
+
+		Diagnostic diag{};
+		diag.range.start.line = startPosition.line;
+		diag.range.start.column = startPosition.column;
+		diag.range.end.line = endPosition.line;
+		diag.range.end.column = endPosition.column;
+		diag.message = message.primary.message;
+		diag.source = "solc";
+		diag.severity = severity;
+
+		for (SourceReference const& secondary: message.secondary)
+		{
+			auto related = DiagnosticRelatedInformation{};
+
+			related.message = secondary.message;
+			related.location.uri = "file://" + secondary.sourceName; // is the sourceName always a fully qualified path?
+			related.location.range.start.line = secondary.position.line;
+			related.location.range.start.column = secondary.startColumn;
+			related.location.range.end.line = secondary.position.line; // what about multiline?
+			related.location.range.end.column = secondary.endColumn;
+
+			diag.relatedInformation.emplace_back(move(related));
+		}
+
+		if (message.errorId.has_value())
+			diag.code = message.errorId.value().error;
+
+		params.diagnostics.emplace_back(move(diag));
+	}
+
+	// some additional analysis (as proof of concept)
+#if 1
+	for (size_t pos = _file.contentString().find("FIXME", 0); pos != string::npos; pos = _file.contentString().find("FIXME", pos + 1))
+	{
+		Diagnostic diag{};
+		diag.message = "Hello, FIXME's should be fixed.";
+		diag.range.start = _file.buffer().toPosition(pos);
+		diag.range.end = {diag.range.start.line, diag.range.start.column + 5};
+		diag.severity = DiagnosticSeverity::Error;
+		diag.source = "solc";
+		params.diagnostics.emplace_back(diag);
+	}
+
+	for (size_t pos = _file.contentString().find("TODO", 0); pos != string::npos; pos = _file.contentString().find("FIXME", pos + 1))
+	{
+		Diagnostic diag{};
+		diag.message = "Please remember to create a ticket on GitHub for that.";
+		diag.range.start = _file.buffer().toPosition(pos);
+		diag.range.end = {diag.range.start.line, diag.range.start.column + 5};
+		diag.severity = DiagnosticSeverity::Hint;
+		diag.source = "solc";
+		params.diagnostics.emplace_back(diag);
+	}
+#endif
+
+	_result.emplace_back(params);
+}
+
+frontend::ASTNode const* LanguageServer::findASTNode(lsp::Position const& _position, std::string const& _fileName)
+{
+	if (!m_compilerStack)
+		return nullptr;
+
+	frontend::ASTNode const& sourceUnit = m_compilerStack->ast(_fileName);
+	auto const sourcePos = sourceUnit.location().source->translateLineColumnToPosition(_position.line + 1, _position.column + 1);
```

Comment:
Shouldn't the `+1` be done inside the routine?

Reply:
Or is there a distinction between "line/column internal" (starts with 0) and "line/column user" (starts with 1)?

Reply:
line/column in solidity starts with 1 whereas line-column in LSP starts with 0. so I'd like not to move this into CharStream, which doesn't know about lsp.